metsim/metsim.py

```diff
@@ -42,71 +47,143 @@ def __init__(self, params:dict):
         Kicks off the disaggregation and queues up data for IO
         """
         # Where we will store the results for writing out
         out_dict = {}
         method = MetSim.methods[MetSim.params.get('method', 'mtclim')]
 
-        # Coerce the forcings into a list if it's not one
-        if type(forcings) is not list:
-            forcings = [forcings]
-
-        # Do the forcing generation and dissaggregation if required
-        for forcing in forcings:
-            met_data = self.read(forcing)
-            for i in range(len(met_data.lat)):
-                out_dict[i] = {}
-                for j in range(len(met_data.lon)):
-                    out_dict[i][j] = method.run(met_data.isel(lat=[i], lon=[j])
-                                                .to_dataframe(), MetSim.params,
-                                                disagg=True)
-            # Write out
-            self.write(out_dict)
-
-        if MetSim.params.get('out_format', 'netcdf') == 'netcdf':
-            self.output.close()
+        # Do the forcing generation and disaggregation if required
+        for i, j in locations:
+            print("Processing {} {}".format(i, j))
+            out_dict["{}_{}".format(i,j)]  = (
+                    method.run(met_data.sel(lat=[i], lon=[j]).to_dataframe(),
+                               MetSim.params, disagg=True))
+        queue.put(out_dict)
 
 
     def find_elevation(self, lat: float, lon: float) -> float:
         """ Use the domain file to get the elevation """
-        lat_idx = np.abs(self.lats - lat).argmin()
-        lon_idx = np.abs(self.lons - lon).argmin()
-        return self.elev[lat_idx, lon_idx]
+        lat_idx = np.abs(self.domain_lats - lat).argmin()
+        lon_idx = np.abs(self.domain_lons - lon).argmin()
+        return self.domain['elev'][lat_idx, lon_idx]
 
 
     def update(self, new_params):
         """Updates the global parameters dictionary"""
         MetSim.params.update(new_params)
 
 
-    def write(self, data: dict):
-        """
-        Dispatch to the right function based on the configuration given
-        """
-        dispatch = {
-                'netcdf' : self.write_netcdf,
-                'ascii'  : self.write_ascii
-                }
-        dispatch[MetSim.params.get('out_format', 'netcdf').lower()](data)
+    def netdf_in_preprocess(self, filename):
+        """Get the extent and load the data"""
+        # Get the information for splitting up the job
+        in_forcing = Dataset(filename, 'r')
+        self.locations = list(itertools.product(in_forcing['lat'], in_forcing['lon']))
+        self.lats = np.array(in_forcing['lat'])
+        self.lons = np.array(in_forcing['lon'])
+        self.lat_idx = {str(lat): i for i, lat in enumerate(self.lats)}
+        self.lon_idx = {str(lon): j for j, lon in enumerate(self.lons)}
+        in_forcing.close()
+        # Now read in the file properly
+        met_data = self.read(filename)
+        return met_data
```

Comment:
Modify this to only return the active grid cells. Keep the extent to write out empty grid cells as `_FillValue`